LibOS/shim/include/shim_vma.h

```diff
@@ -90,39 +87,59 @@ static inline PAL_FLG PAL_PROT (int prot, int flags)
 int init_vma (void);
 
 /* Bookkeeping mmap() system call */
-int bkeep_mmap (void * addr, uint64_t length, int prot, int flags,
-                struct shim_handle * file, uint64_t offset, const char * comment);
+int bkeep_mmap (void * addr, uint64_t length,
+                int prot, int flags,
+                struct shim_handle * file, uint64_t offset,
+                const char * comment);
 
 /* Bookkeeping munmap() system call */
 int bkeep_munmap (void * addr, uint64_t length, int flags);
 
 /* Bookkeeping mprotect() system call */
 int bkeep_mprotect (void * addr, uint64_t length, int prot, int flags);
 
-/* Get vma bookkeeping handle */
-void get_vma (struct shim_vma * vma);
-void put_vma (struct shim_vma * vma);
+/* Looking up VMA that contains [addr, length) */
+int lookup_vma (void * addr, struct shim_vma_val * vma);
 
-/* Returns 0 on success, -E* on failure.
-   Calls `get_vma` on the result before returning it.
-*/
-int lookup_supervma (const void * addr, uint64_t len, struct shim_vma ** vma);
-int lookup_overlap_vma (const void * addr, uint64_t len, struct shim_vma ** vma);
+/* Looking up VMA that overlaps with [addr, length) */
+int lookup_overlap_vma (void * addr, uint64_t length,
+                        struct shim_vma_val * vma);
 
-struct shim_vma * next_vma (struct shim_vma * vma);
-
-void * get_unmapped_vma (uint64_t len, int flags);
-void * get_unmapped_vma_for_cp (uint64_t len);
+/*
+ * Looking for an unmapped space and then adding the corresponding bookkeeping
+ * (more info in bookkeep/shim_vma.c).
+ *
+ * Note: the first argument is "top_addr" because the search is top-down.
+ */
+void * bkeep_unmapped (void * top_addr, void * bottom_addr, uint64_t length,
+                       int prot, int flags, struct shim_handle * file,
+                       uint64_t offset, const char * comment);
+
+static inline void *
+bkeep_unmapped_any (uint64_t length, int prot, int flags,
+                    struct shim_handle * file, uint64_t offset,
+                    const char * comment)
+{
+    return bkeep_unmapped(PAL_CB(user_address.end),
+                          PAL_CB(user_address.start),
+                          length, prot, flags, file, offset, comment);
+}
 
-int dump_all_vmas (struct shim_thread * thread, char * buf, uint64_t size);
+void * bkeep_unmapped_heap (uint64_t length, int prot, int flags,
+                            struct shim_handle * file, uint64_t offset,
+                            const char * comment);
 
-void unmap_all_vmas (void);
+/*
+ * Dumping all *non-internal* VMAs into a user-allocated buffer ("size" must be
+ * number of entries in the buffer). Return number of filled entries if
+ * succeeded, or -EOVERFLOW if the buffer is too small.
+ */
+int dump_all_vmas (struct shim_vma_val * vmas, size_t size);
 
 /* Debugging */
 void debug_print_vma_list (void);
```

Comment:
There's no implementation for this function, please remove (or restore the implementation, I'm not sure if it's needed).